allennlp/training/initializers.py

```diff
@@ -0,0 +1,117 @@
+from typing import Dict, Callable
+import logging
+import re
+
+import torch.nn.init
+
+from allennlp.common.params import Params
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+class InitializerApplicator:
+    """
+    Applies initializers to the parameters of a Module based on regex matches.
+    All parameters in the Module will be initialized.
+
+    """
+    def __init__(self, all_initializers: Dict[str, Callable[[torch.Tensor], None]] = None) -> None:
+        """
+        Parameters
+        ----------
+        all_initializers : Dict[str, Callable[[torch.Tensor], None]], optional (default = None)
+            A dictionary mapping parameter regexes to initializers to be applied to parameters
+            matching the regex. Additionally, it may include a "default" field containing a
+            initializer, which will be used in the case that the Applicator encounters a parameter
+            which does not match any of the regexes provided.
+
+        """
+        self._initializers = all_initializers or {}
+        self._default_initializer = self._initializers.pop("default", torch.nn.init.normal)
+
+    def __call__(self, module: torch.nn.Module) -> None:
+        """
+        Applies a series of initializers to all parameters in a module if those parameters match a
+        regex. If no explicitly specified initializers are applied, a default initializer is applied.
+
+        Parameters
+        ----------
+        module : torch.nn.Module, required.
+            The Pytorch module to apply the initializers to.
+        """
+        # Store which initialisers were applied to which parameters.
+        not_explicitly_initialized_parameters = []
+        for name, parameter in module.named_parameters():
+            is_initialized = False
+            for initializer_regex, initializer in self._initializers.items():
+                if self.parameter_name_matches_regex(initializer_regex, name):
+                    initializer(parameter)
+                    logger.info("Initializing %s using %s "
+                                "Intitializer.", name, initializer_regex)
+                    is_initialized = True
```

Comment:
You could consider using the ["for ... else"](https://stackoverflow.com/questions/9979970/why-does-python-use-else-after-for-and-while-loops/9980752#9980752) pattern here, instead of having a sentinel.